### ISSUE TITLE
Upgrade react-navigation to v3

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3649]
Forking this out from #3573. Probably I should have made separate issues in the first place.

We use `react-navigation` for navigation in the app. We're on version 2.18.3. There's a new major version out, v3. (We were on v1 until not long ago; getting to v2 was #3573.)

The v3 releases contain various bugfixes, as well as compatibility with newer other software. Issues blocked on this upgrade include #3569, and possibly #2716 and #3566.

There's partial work toward this issue in #3502.

**Comment by gnprice** [comment]
Release notes, since we don't already have them linked here:
https://reactnavigation.org/blog/2018/11/17/react-navigation-3.0.html

Note that installation has gotten a bit more complicated:
https://reactnavigation.org/docs/en/3.x/getting-started.html#installation
Though not much so if on RN v0.60+ -- so if we do this after #3548 it'll be somewhat simpler.

**Comment by awasthishubh** [comment]
@gnprice I would like to do this. Can I?

**Comment by gnprice** [comment]
@awasthishubh I've just assigned you #3700, which you claimed there; let's start with that. :slightly_smiling_face:

**Comment by Maskedman99** [comment]
```react-navigation``` is currently at v5.x and they've changed the package names
```
    react-navigation -> @react-navigation/native
    react-navigation-stack -> @react-navigation/stack
    react-navigation-tabs -> @react-navigation/bottom-tabs, @react-navigation/material-top-tabs  
    react-navigation-material-bottom-tabs -> @react-navigation/material-bottom-tabs
    react-navigation-drawer -> @react-navigation/drawer
```
See https://reactnavigation.org/docs/upgrading-from-4.x

**Comment by chrisbobbe** [comment]
It appears that FlowTyped does have a [libdef](https://github.com/flow-typed/flow-typed/tree/master/definitions/npm/react-navigation_v3.x.x) for this version! So we very likely won't have to use Flowgen for this upgrade.

**Comment by chrisbobbe** [comment]
Posting some details on my procedure that seem too long to fit into the main commit message:

# Update / Add Deps

We start by just `yarn upgrade`ing react-navigation to ^3.13.0 (the latest 3.x).

With that, we get the following peer-dep output:

```
warning " > react-navigation-tabs@0.8.4" has incorrect peer dependency "react-navigation@^2.0.0".
warning "react-navigation > react-navigation-stack@1.5.3" has unmet peer dependency "react-native-gesture-handler@^1.0.0".
warning "react-navigation > react-navigation-drawer@1.4.0" has unmet peer dependency "react-native-gesture-handler@^1.0.12".
warning "react-navigation > @react-navigation/native@3.6.5" has unmet peer dependency "react-native-gesture-handler@*".
```

We upgrade react-navigation-tabs to ~1.2.0, which we choose because our new react-navigation v3 has that as a direct dependency (more on this below).

The react-navigation v4.x [sic] upgrade guide [3] recommends starting *that* upgrade by adding three direct dependencies, for an interesting reason:

- react-navigation-stack@^1.7.3
- react-navigation-tabs@^1.2.0 (note we didn't use this version range; see below)
- react-navigation-drawer@^1.4.0

That reason is,

"""
This will install the versions compatible with your code if you were using `react-navigation@3.x`, so you wouldn't need any more changes beyond changing the imports.
"""

So it seems like we'll want (or need) to depend on these three directly, in react-navigation v4, probably because they'll newly be peer-deps of react-navigation.

Might as well keep react-navigation-tabs (and close #4114), update its version, and add the other two right now. However, the indicated version ranges aren't *quite* right for v3 compatibility; they would resolve react-navigation-stack to 1.10.3, which appears to require react-navigation 4.x (expressed as a peer-dep). To be safe, we just pick the version ranges straight from react-navigation v3's direct dependencies:

- react-navigation-drawer@~1.4.0
- react-navigation-stack@1.5.3
- react-navigation-tabs@~1.2.0

At this point, the peer-dep output is the following:

```
warning "react-navigation > @react-navigation/native@3.6.5" has unmet peer dependency "react-native-gesture-handler@*".
warning " > react-navigation-drawer@1.4.0" has unmet peer dependency "@react-navigation/core@*".
warning " > react-navigation-drawer@1.4.0" has unmet peer dependency "@react-navigation/native@*".
warning " > react-navigation-drawer@1.4.0" has unmet peer dependency "react-native-gesture-handler@^1.0.12".
warning " > react-navigation-stack@1.5.3" has unmet peer dependency "@react-navigation/core@^3.5.0".
warning " > react-navigation-stack@1.5.3" has unmet peer dependency "@react-navigation/native@^3.6.2".
warning " > react-navigation-stack@1.5.3" has unmet peer dependency "react-native-gesture-handler@^1.0.0".
```

For each package name, we take the intersection of its valid ranges, adding

- @react-navigation/core@^3.5.0
- @react-navigation/native@^3.6.2
- react-native-gesture-handler@^1.0.12

at which point there are no more peer-dep warnings.

The 3.x installation guide also says we need to add `react-native-reanimated` (we've seen this once before, in unrelated work [4]). But I don't think we need to yet, and in particular, I'd rather not add it until we know what version of it we want:

- No peer-dep warning is telling us to use it.
- `react-navigation-tabs` doesn't import it until react-navigation/tabs@e6efae638, which is tags/v2.0.0-alpha.0~3. `yarn why` shows we've resolved to use react-navigation-tabs@1.2.0.
- `react-navigation-drawer` doesn't import it until react-navigation/drawer@ec5e99cd4, which is tags/v2.0.0-alpha.0~7). `yarn why` shows we've resolved to use react-navigation-drawer@1.4.0.
- `react-navigation-stack` doesn't import it until react-navigation/stack@40a33b3fb, which is tags/v2.0.0-alpha.6~64). `yarn why` shows we've resolved to use react-navigation-stack@1.5.3.
- The above three, plus `@react-navigation/core` and `@react-navigation/native`, are the only things listed under `dependencies` in v3 of `react-navigation`. The latter two don't use `react-native-reanimated` at any version.
- (`react-navigation-animated-switch` is advertised as making use of `react-native-reanimated`; naturally, that project starts importing it at its initial commit. But `yarn why` shows that we're not using `react-navigation-animated-switch` at all.)

Next, the doc says we need to add some Java to our MainActivity.java, to get react-native-gesture-handler up and running. Since bdce7d0a9, we don't have a MainActivity.java; ours was converted to Kotlin. The react-native-gesture-handler setup doc [5] doesn't have a Kotlin version of the added code either, so I temporarily revived MainActivity.java from bdce7d0a9^, added it in there, used Android Studio to convert the file to Kotlin, and inserted the translated new Kotlin into our MainActivity.kt.

At this point, Jest fails because react-native-gesture-handler depends on a native module that won't be available in Jest. So, follow a recommendation [6] to run that package's own Jest setup, as a fix. (I wish more packages shipped with these and suggested this approach!)

Even though we're trying to stop using react-navigation-redux-helpers (and we'll need to by react-navigation v5; see #3804), we haven't yet. And we have to take its v3 and handle the breaking changes [7]. All three breaking changes apply to us, but they're small and simple. Also, removing this library from our Flow config's `[ignore]`s seems to be fine, and gets us more type-checking coverage. There's actually a known gap in type-checking, though; see the [v3.0.8 release notes](https://github.com/react-navigation/redux-helpers/releases/tag/3.0.8), which suggests that `react-navigation-redux-helpers`'s types will be incomplete without the @react-navigation/core libdef. I tried adding that, and it clashes with the react-navigation v3 libdef in a few places, and decided not to. I figure we'd rather not have to keep tweaking two libdefs to keep in sync with each other. Also:
- Where the two disagreed, it seemed like `react-navigation_v3.x.x` was out-of-date, but in multiple places, those issues get fixed with `react-navigation_v4.x.x`
- @react-navigation/core [describes itself](https://github.com/react-navigation/react-navigation/tree/main/packages/core#react-navigationcore) as a thing for "building navigators independent of the platform". We're not in the business of building navigators; we *use* ones (like react-navigation-tabs) that are already built. Given this, it seems unlikely that we'll reasonably need a libdef for it
- As mentioned earlier, we want to move away from using react-navigation-redux-helpers ASAP.

Also, run `yarn yarn-deduplicate && yarn` as prompted by `tools/test deps`.

# Libdefs

A react-navigation v3 libdef from FlowTyped exists, and it looks good! This saves us hours of work. Particularly helpful and admirable is the fact that they've gone and painstakingly copied some things over from other modules (with a few minor errors, some or all of which will be fixed in v4; see above), since they can't be imported [8]. There is a small hiccup (not that libdef's fault); `react-navigation-tabs` shows some dubious errors that I traced back that library's built-in types being badly written. I found that they go away with the built-in types in v2.4.1, but peer-dep warnings tell us that we can't use that version until we're on react-navigation v4 (#4248). Ah, well; add a temporary line to our Flow config's `[ignore]`s.

For the other dependencies added in this commit:
- There are libdefs for @react-navigation/core (mentioned above) and @react-navigation/native, but since we don't directly consume anything from these, we don't expect to need the libdefs, and at least the @react-navigation/core libdef at this version conflicts with the react-navigation libdef.
- We don't directly consume anything in react-native-gesture-handler in our JS.
- There's a libdef for react-navigation-tabs, but for "2.x.x"; we're not there yet. We'll be there soon, with react-navigation v4, though.
- There's a libdef for react-navigation-stack for "1.x.x", great! Add that.
- There's a libdef for react-navigation-drawer, but for "2.x.x"; we're not there yet. We'll be there soon, with react-navigation v4, though.

Now all we have to do is handle the announced breaking changes [2], as follows:

- "Explicit app container required for the root navigator"
  - Easy enough; make sure our `AppNavigator` is wrapped with the appropriate function. Consulting the new version of react-navigation-redux-helpers, they'd actually like us to use something they provide instead, `createReduxContainer`, so that that library's functionality (which we hope to stop using in #3804) still works. Sure.
  - This seems associated with a breakage we found a workaround for; see [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Dynamic.20routes.20in.20react-navigation/near/1008268)
- "Renamed navigationOptions in navigator configuration"
  - This doesn't seem to apply to us; when we use `navigationOptions`, we always seem to do so in a way that's specifically intended for the corresponding navigator. We don't set `navigationOptions` intending for the settings to cascade as defaults to descendent navigators, which was possible. That cascading is still possible, you just have to explicitly use `defaultNavigationOptions` if you want it. Seems much clearer this way.
- "Drawer now keeps inactive tabs in memory by default"
  - We're not using drawer navigators.
- "Default stack background color is now white"
  - In our one call site of `createStackNavigator`, we already set the background color to white.

[1] https://reactnavigation.org/docs/3.x/getting-started/
[2] https://reactnavigation.org/blog/2018/11/17/react-navigation-3.0/
[3] https://reactnavigation.org/docs/4.x/upgrading-from-3.x/#install-the-new-packages
[4] https://github.com/zulip/zulip-mobile/issues/4217#issue-673154603
[5] https://docs.swmansion.com/react-native-gesture-handler/docs/#installation
[6] https://github.com/software-mansion/react-native-gesture-handler/issues/344#issuecomment-489547513
[7] https://github.com/react-navigation/redux-helpers/releases/tag/3.0.0
[8] https://github.com/zulip/zulip-mobile/issues/3458#issuecomment-639859987
